src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProjectFactory.cs

```diff
@@ -779,49 +796,79 @@ private static bool CanConvertMetadataReferenceToProjectReference(Solution solut
     /// Gets or creates a PortableExecutableReference instance for the given file path and properties.
     /// Calls to this are expected to be serialized by the caller.
     /// </summary>
-    public static PortableExecutableReference CreateMetadataReference_NoLock(string fullFilePath, MetadataReferenceProperties properties, SolutionServices solutionServices)
+    public static PortableExecutableReference CreateMetadataReference_NoLock(
+        string fullFilePath, MetadataReferenceProperties properties, SolutionServices solutionServices)
     {
-        var reference = solutionServices.GetRequiredService<IMetadataService>().GetReference(fullFilePath, properties);
-        return reference;
+        return solutionServices.GetRequiredService<IMetadataService>().GetReference(fullFilePath, properties);
     }
 
     private void StartRefreshingMetadataReferencesForFile(object? sender, string fullFilePath)
-    {
-        var asyncToken = WorkspaceListener.BeginAsyncOperation(nameof(StartRefreshingMetadataReferencesForFile));
+        => StartRefreshingReferencesForFile(
+            fullFilePath,
+            static project => project.MetadataReferences.OfType<PortableExecutableReference>(),
+            static reference => reference.FilePath!,
+            static (@this, reference) => CreateMetadataReference_NoLock(reference.FilePath!, reference.Properties, @this.SolutionServices),
+            static (projectUpdateState, oldReference, newReference) => projectUpdateState
```

Comment:
nit: I see all these lambdas together and wonder if named arguments would improve the readability.

Reply:
@JoeRobich sure.  I can add that :)